test/external-modules/error-query/src/main/java/org/elasticsearch/test/errorquery/ErrorQueryBuilder.java

```diff
@@ -133,4 +166,80 @@ protected int doHashCode() {
     public TransportVersion getMinimalSupportedVersion() {
         return TransportVersion.ZERO;
     }
+
+    static void sleep(long millis) {
+        try {
+            Thread.sleep(millis);
+        } catch (InterruptedException e) {
+            // ignore
+        }
+    }
+
+    /**
+     * ErrorQuery uses MatchAllDocsQuery when doing searches.
+     * It can optionally add warnings, throw exceptions and sleep for specified "stall times"
+     * based on the information in the provided IndexError class.
+     */
+    static class ErrorQuery extends Query {
+        private final IndexError indexError;
+        private volatile boolean sleepCompleted;
+        private final MatchAllDocsQuery matchAllQuery;
+
+        ErrorQuery(IndexError error, SearchExecutionContext context) {
+            this.indexError = error;
+            this.sleepCompleted = false;
+            this.matchAllQuery = new MatchAllDocsQuery();
+
+            if (error.getShardIds() != null) {
+                boolean match = false;
+                for (int shardId : error.getShardIds()) {
+                    if (context.getShardId() == shardId) {
+                        match = true;
+                        break;
+                    }
+                }
+                if (match == false) {
+                    return;
+                }
+            }
+            final String header = "[" + context.index().getName() + "][" + context.getShardId() + "]";
+            if (error.getErrorType() == IndexError.ERROR_TYPE.WARNING) {
+                HeaderWarning.addWarning(header + " " + error.getMessage());
+            } else if (error.getErrorType() == IndexError.ERROR_TYPE.EXCEPTION) {
+                if (indexError.getStallTimeSeconds() > 0) {
+                    sleep(indexError.getStallTimeSeconds() * 1000L);
+                }
+                throw new RuntimeException(header + " " + error.getMessage());
+            }
+        }
+
+        @Override
+        public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) {
+            if (indexError.getStallTimeSeconds() > 0 && sleepCompleted == false) {
+                sleep(indexError.getStallTimeSeconds() * 1000L);
+                sleepCompleted = true;
```

Comment:
Have you found that we're recreating the Weight multiple times during a search?  I can see it happening in separate phases but I think I'd expect those to be recreated from the QueryBuilder each time.

Reply:
Yeah, in one manual test run I did it was stalling for twice the time I specified, so I added logging and saw that createWeight was getting called twice.

> I think I'd expect those to be recreated from the QueryBuilder each time.

Oh, interesting. Let me do some additional testing then. Perhaps I track "sleepTimeAmount" in a shared IndexError object instead that is passed into each Query object. Would that work? Are QueryBuilder's also recreated in the separate search phases?

Reply:
I suppose you could also argue that this is only every going to be a lower bound on the sleep time (because the CPU might be busy, or you get a GC pause, or whatever) so taking twice as long as asked doesn't matter that much.  Also, it's a test-only class so it's not as if any consumers are going to be relying on it?

Reply:
On further analysis, the "double call" to ErrorQuery.createWeight only happens if you aren't throwing an Exception and do a full search.

The stack trace for the first call shows it is being called from `QueryPhase.searchWithCollectorManager`

```
	at org.elasticsearch.test.errorquery.ErrorQueryBuilder$ErrorQuery.createWeight(ErrorQueryBuilder.java:223)
	at org.apache.lucene.core@9.7.0/org.apache.lucene.search.IndexSearcher.createWeight(IndexSearcher.java:904)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.internal.ContextIndexSearcher.createWeight(ContextIndexSearcher.java:241)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:327)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.query.QueryPhase.searchWithCollectorManager(QueryPhase.java:285)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.query.QueryPhase.addCollectorsAndSearch(QueryPhase.java:252)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.query.QueryPhase.executeQuery(QueryPhase.java:141)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:68)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:495)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:647)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:520)
```

whereas the second call (where `sleepComplete=true`) is happening during the aggregation sub-phase:

```
	at org.elasticsearch.test.errorquery.ErrorQueryBuilder$ErrorQuery.createWeight(ErrorQueryBuilder.java:223)
	at org.apache.lucene.core@9.7.0/org.apache.lucene.search.IndexSearcher.createWeight(IndexSearcher.java:904)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.internal.ContextIndexSearcher.createWeight(ContextIndexSearcher.java:241)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.aggregations.bucket.filter.QueryToFilterAdapter.weight(QueryToFilterAdapter.java:245)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.aggregations.bucket.filter.QueryToFilterAdapter.count(QueryToFilterAdapter.java:200)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.aggregations.bucket.filter.FilterByFilterAggregator.collectCount(FilterByFilterAggregator.java:270)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.aggregations.bucket.filter.FilterByFilterAggregator.getLeafCollector(FilterByFilterAggregator.java:253)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.aggregations.AggregatorBase.getLeafCollector(AggregatorBase.java:220)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.aggregations.AdaptingAggregator.getLeafCollector(AdaptingAggregator.java:85)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.aggregations.MultiBucketCollector$1.getLeafCollector(MultiBucketCollector.java:92)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.aggregations.BucketCollector$BucketCollectorWrapper.getLeafCollector(BucketCollector.java:86)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.query.QueryPhaseCollector.getLeafCollector(QueryPhaseCollector.java:163)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.internal.ContextIndexSearcher.searchLeaf(ContextIndexSearcher.java:440)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:404)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:341)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:332)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.query.QueryPhase.searchWithCollectorManager(QueryPhase.java:285)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.query.QueryPhase.addCollectorsAndSearch(QueryPhase.java:252)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.query.QueryPhase.executeQuery(QueryPhase.java:141)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:68)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:495)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:647)
	at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:520)
```

so based on this it looks safe to keep the `sleepCompleted` state in the ErrorQuery itself, since it is being used twice in the same QueryPhase.

Reply:
Ah right, yes that makes sense if it's doing a filter-by-filter aggregation as well.